VRSC sapling compatibility for Iguana dpow and splitfunds

DeckerSU opened the pull request:
This PR implements:
- sapling (version=4) txes for dpow
- sapling (version=4) txes for splitfunds
- changes applied only to VRSC and VRSCTEST coin symbols
- fixed bug with coins which have len(symbol) > 7 - https://github.com/DeckerSU/SuperNET/commit/a282cf96847621ca0ff71cff79249913d37fdaf8
- now iguana uses additional dependency `libsodium` for calc BLAKE2b-256 hashes - https://github.com/DeckerSU/SuperNET/commit/fc33878df9dc8d97b50919adaac063aec5fa9895

**NB!** We should merge this PR with `dev` branch only after VRSC sapling activation, i.e. **227520 block**, to avoid produce txes with incorrect version for current VRSC mainchain.

DeckerSU committed:
fix safecopy: %s too long 8 error on RPC with coins len(name) > 7

Added symbol_len in iguana_coinchoose. Here was a sizeof(symbol)
in safecopy call args, but sizeof(char *) always = 8 on 64-bit systems,
as a result if symbol have more than 7 letters, like "VRSCTEST"
safecopy returns an error.

DeckerSU committed:
VRSC iguana sapling changes [part 1]

- already implemented unsigned raw tx creation (thx @artemii235 for mm codebase)
  tested on iguana_utxoduplicates (dpow is still untested)
- todo: bitcoin_sigtxid implement signing for overwintered & sapling
- todo: check bitcoin_verifyvins
- todo: remove debug printouts

DeckerSU committed:
+ libsodium.lib for Windows, iguana needed BLAKE2b-256 calc for preimage

before build iguana under Windows you should download and compile
libsodium.lib & libsodium.dll by using iguana_build_depends.cmd .
in marketmaker this accomplished by using cmake, but iguana didn't use
cmake build.

DeckerSU committed:
VRSC iguana sapling changes [part 2]

- bitcoin_sigtxid sapling tx sighash preimage changes (thx to @artemii235)

now splitfunds for VRSC and VRSCTEST produce correct signed txes.

DeckerSU committed:
- remove debug printouts

DeckerSU committed:
+ iguana_rwjoinsplit: add proof_size param

DeckerSU committed:
+ test build iguana for linux (added libsodium)

on ubuntu-like systems libsodium can be easily installed by
`sudo apt install libsodium-dev`, but we can make download and
compile it from sources in additional bash script.

DeckerSU committed:
dpow_notarytx: start preparations for VRSC sapling

usesigs? when we should create unsignedrawtx and when preimage for signing?

DeckerSU committed:
dpow_notarytx: correct sapling unsigned raw tx with usesigs=0

now we have correct format of unsigned rawtx with sapling support inside
dpow_notarytx and correct format of signed raw tx when usesigs=0.
but as we don't yet create sapling preimage - sigs are still incorrect,
and notary tx can't be broadcasted.

DeckerSU committed:
+ ISNOTARYNODE and LIQUIDITY_PROVIDER for Windows build

DeckerSU committed:
+ add debug printout when signing vins

DeckerSU committed:
+ amount field in dpow_vins

dpow_signrawtransaction used signrawtransaction RPC call to daemon
when obtain scriptSig for known notary vin. but signrawtransaction
RPC call in VRSC with activated sapling need to pass amount as required
parameter, otherwise it produced incorrect sig. spent many time
to understand this.

DeckerSU committed:
remove debug printouts

DeckerSU committed:
+ lsodium to linker

now iguana uses additional depependency libsodium (!) for calculating
BLAKE2b-256 hash in preimage, according to ZIP-0243
(https://github.com/zcash/zips/blob/master/zip-0243.rst). On Ubuntu
based systems it can be easily installed via `sudo apt install libsodium-dev` .

jl777 commented:
this looks fantastic!
however could we make a new field in struct iguana_info like int32_t sapling
then instead of checks against VRSC, it can just use coin->sapling, or even height > coin->sapling to enable height activation.

you can have the VRSC specific coin->sapling = <activation height> in the iguana_setcoin() function, or it could be something that it gets from the JSON so that no code changes would be needed to activate sapling for other coins.

DeckerSU committed:
sapling param added to json

- added sapling param in `struct iguana_info`
- each coin now have coin->sapling param (sapling=0 - sapling disabled, sapling !=0 - sapling enabled), we can set "sapling":1 in JSON when calling "addcoin" method.
- removed hardcoded "VRSC" and "VRSCTEST", now we should use "sapling" param in "addcoin" method
- removed unused zcash param in bitcoin_sigtxid

DeckerSU committed:
Merge branch 'dev-decker-vrsc' of ssh://github.com/DeckerSU/SuperNET into dev-decker-vrsc

DeckerSU commented:
Done - https://github.com/jl777/SuperNET/pull/1001/commits/739b67f0a64bb6208c6e08c8bb93307fec98f335 .
